### PR TITLE
Update Prompts | Prompts [2/X]

### PR DESCRIPTION
**Pull request by joycrwu**
## Overview

Implemented endpoints for updating the contents of a prompt. 

## Changes Made

Made a new controller for updating the question name and the question placeholder of an existing prompt. 

## Test Coverage

Manual testing via Postman
Endpoint: http://127.0.0.1:8000/api/prompts/<int: prompt_id> POST
**Example Request Body:**

 ```
{
    "question_name" : "Favorite city",
    "question_placeholder" : "Tokyo"
}
```

**Example Response:**

 ```
{
    "id": 1,
    "question_name": "Favorite city",
    "question_placeholder": "Tokyo"
}
```

## Related PRs or Issues (delete if not applicable)
#23 
#22

### CODE REVIEW

File: src/api/utils.py [hunk 1]
```diff
@@ -24,3 +24,12 @@ def update(model, attr_name, attr_value):
         and attr_value != getattr(model, attr_name)
     ):
         setattr(model, attr_name, attr_value)
+
+def modify_attribute(model, attr_name, attr_value):
```
Comment (chalo2000) [inline src/api/utils.py]: this should be unnecessary to use and you can just modify the attribute directly

File: src/prompts/controllers/create_prompt_controller.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def __init__(self, data, serializer):
+        self._data = data
+        self._serializer = serializer
+
+    def process(self):
+        name = self._data.get("question_name")
+        if name is None:
+            return failure_response("No name provided.", status.HTTP_400_BAD_REQUEST)
+        
+        label = self._data.get("question_placeholder")
```
Comment (chalo2000) [inline src/prompts/controllers/create_prompt_controller.py]: fix this as well as in the last PR

File: src/prompts/controllers/update_prompt_controller.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        if not prompt:
+            return failure_response("Prompt does not exist", status.HTTP_404_NOT_FOUND)
+        prompt = prompt[0]
+
+        # Extract attributes
+        name = self._data.get("question_name")
+        label = self._data.get("question_placeholder")
+
+        # Modify new fields
+        modify_attribute(prompt, "question_name", name)
```
Comment (chalo2000) [inline src/prompts/controllers/update_prompt_controller.py]: directly modify when you remove the helper function by doing `prompt["question_name"] = name` which is what the helper was doing in an unnecessarily complicated way

File: src/prompts/views.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    def get(self, request):
+        """Get all prompts."""
+        prompts = Prompt.objects.all()
+        return success_response(
+            self.serializer_class(prompts, many=True).data, status.HTTP_200_OK
+        )
+
+    def post(self, request):
+        """Create a prompt."""
+        try:
```
Comment (chalo2000) [inline src/prompts/views.py]: remove try except

File: src/prompts/views.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        prompt = Prompt.objects.filter(id=id)
+        if not prompt:
+            return failure_response("Prompt does not exist", status.HTTP_404_NOT_FOUND)
+        return success_response(
+            self.serializer_class(prompt[0]).data, status.HTTP_200_OK
+        )
+
+    def post(self, request, id):
+        """Update prompt by id."""
+        try:
```
Comment (chalo2000) [inline src/prompts/views.py]: ditto

**Review by chalo2000 — APPROVED** [review]
LGTM

File: src/api/utils.py (hunk 1 above)
Comment (joycrwu) [reply src/api/utils.py]: it wouldn't let me access it directly, but i saw that we already had another update function that does what modify_attribute did

File: src/prompts/controllers/create_prompt_controller.py (hunk 2 above)
Comment (joycrwu) [reply src/prompts/controllers/create_prompt_controller.py]: fixed

File: src/prompts/controllers/update_prompt_controller.py (hunk 3 above)
Comment (joycrwu) [reply src/prompts/controllers/update_prompt_controller.py]: same as above comment

File: src/prompts/views.py (hunk 4 above)
Comment (joycrwu) [reply src/prompts/views.py]: removed

File: src/prompts/views.py (hunk 5 above)
Comment (joycrwu) [reply src/prompts/views.py]: ditto

**Review by katesliang — APPROVED** [review]
lgtm

**Pull request merged**